[최지수] Step5 리뷰 요청합니다~

jisu1211 opened the pull request:
안녕하세요 우빈님~!
4단계 피드백 수정하고 5단계도 구현 완료하여 리뷰 요청합니다~
우빈님이 세심하게 피드백해주시고 가이드해주신 덕분에 1주일이라는 짧았던 시간이었지만,
배워가는 것은 정말 많은 것 같습니다! 
앞으로도 피드백 해주신 부분들 신경써서 습관으로 만들 수 있으면 좋겠네요! ㅎㅎ

마지막 피드백도 잘 부탁드리고 좋은 하루 보내세요~ㅎㅎ
감사합니다🥳☺️

jisu1211 committed:
refactor: 4단계 피드백 반영
- 클래스, 메소드에 대한 네이밍 변경
- StringExceptionCode enum 추가

jisu1211 committed:
docs: 5단계 README.md 작성

jisu1211 committed:
feat: UserPurchase 클래스 생성
- 사용자의 로또 구매 정보를 관리하는 클래스.
  - 로또의 구매 금액을 보관한다.
  - 로또의 구매 금액은 로또 기본 가격(1000원)보다 커야한다. -> 작은 경우 IllegalArgumentException 발생
  - 로또의 구매 수량을 보관한다.
  - 로또의 수동 구매량을 입력받아 총 구매량과 비교해 수동 구매할 수 있는지 확인한다.
  - 로또의 수동 구매량을 입력받아 자동 구매량을 계산할 수 있다.

jisu1211 committed:
refactor: 기존의 Order 클래스 제거

jisu1211 committed:
feat: UserOrder 기능 구현
- UserOrder
  - 사용자의 로또 주문을 관리하는 클래스.
  - 사용자의 로또 구매정보(UserPurchase)를 가진다.
  - 로또의 수동 구매량을 입력받아 구매 여부를 확인할 수 있다. -> 구매할 수 없는 경우 IllegalArgumentException 발생
  - 로또를 수동 구매/자동 구매로 나누어 주문할 수 있다(통합된 Lottos 리턴).
    - AutoOrder(Order)
      - 로또의 자동 구매를 관리하는 클래스.
      - 입력된 수량만큼 Store에서 자동 구매한다.
    - ManualOrder(Order)
      - 로또의 수동 구매를 관리하는 클래스.
      - 입력된 수량까지 하나씩 Store에서 수동 구매한다.

test: UserOrderTest 추가

jisu1211 committed:
feat: Store에서 Lotto를 auto/manual로 구분해 구매할 수 있도록 구현
- WinningLotto에서 Store 코드를 재사용할 수 있도록 수정
- LottoExceptionCode에 exception code 추가

test: StoreTest 추가

jisu1211 committed:
feat: 로또를 수동/자동으로 나눠서 구매하기 위한 입출력 추가

limwoobin commented on src/main/java/study/lotto/domain/number/LottoGenerator.java:
```diff
@@ -20,15 +20,15 @@ public class CacheLottoNumbers {
         } catch (IllegalArgumentException iae) {
             throw new IllegalArgumentException(LottoExceptionCode.INVALID_LOTTO_NUMBER.getMessage());
         }
     }
```
toLottoNumber가 String, Integer 두 타입을 모두 받고있네요 ㅎㅎ
이 두 방식을 하나의 인자만 받게끔 제한시켜 toLottoNumber를 하나로 사용하는건 어떨까요?

jisu1211 replied:
네 말씀해주신 방향으로 고민해보겠습니다~

limwoobin commented on src/main/java/study/lotto/domain/Store.java:
```diff
@@ -33,18 +33,18 @@ public static List<Lotto> buy(int quantity) {
         return rangeOfLottoNumbers.subList(LOTTO_STAT_IDX, LOTTO_END_IDX)
                 .stream()
-                .map(CacheLottoNumbers::of)
+                .map(LottoGenerator::toLottoNumber)
```
LottoGenerator로 부터 아에 로또객체를 리턴받는게 
더 객체의 역할에 알맞지 않을까요??

jisu1211 replied:
로또 객체를 리턴하는 방향으로 수정해보겠습니다 ㅎㅎ

limwoobin commented on src/main/java/study/lotto/domain/Lotto.java:
```diff
@@ -17,7 +17,7 @@ public Lotto(List<Integer> numbersFromStore) {
     private Set<LottoNumber> createLottoNumbers(List<Integer> numbersFromStore) {
         return numbersFromStore.stream()
-                .map(CacheLottoNumbers::of)
+                .map(LottoGenerator::toLottoNumber)
```
여기도 로또자체로 리턴받는건 어떨까요 ㅎㅎ

limwoobin commented on src/main/java/study/lotto/LottoGames.java:
```diff
@@ -23,11 +25,31 @@ public void start() {
+        userOrder.prepareOrder(manualQuantity);
+        inputManualLottos(manualQuantity, userOrder);
+
+        Lottos lottos = userOrder.order();
```
userOrder라는 곳에서 Lottos를 리턴하는군요
Store라는 객체가 있는데 userOrder에서 로또를 리턴하는게 어색하네요
아에 Store에게 UserOrder와 같은 객체를 전달해서 리턴받는 방법도 있을것같습니다.

jisu1211 replied:
말씀해주신대로 Store에게 UserOrder를 전달한다는게 자연스럽게 느껴지네요~ 수정해보도록 하겠습니다 ㅎㅎ

limwoobin commented on src/main/java/study/lotto/LottoGames.java:
```diff
@@ -23,11 +25,31 @@ public void start() {
+        return lottos;
+    }
+
+    private void inputManualLottos(int manualQuantity, UserOrder userOrder) {
+        if(manualQuantity >= MANUAL_ORDER_MIN) {
+            output.printMessage("수동으로 구매할 번호를 입력해 주세요.");
+            input.inputManualLottos(manualQuantity, userOrder);
+        }
+    }
+
+    private void printOrderResult(String orderResult) {
+        String[] orderMsg = Splitter.split(orderResult);
+        output.printMessage("수동으로 " + orderMsg[0] + "개, 자동으로 " + orderMsg[1] + "개를 구매했습니다.");
     }
```
이런 메서드는 view영역으로 이동시키는게 어떨까요??

limwoobin commented on src/main/java/study/lotto/domain/order/UserOrder.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class UserOrder {
```
UserOrder와 UserPurchase의 차이가 어떻게 될까요??
두 객체를 합쳐도 크게 문제는 없을것같은데 어떻게 생각하시나요??

jisu1211 replied:
UserPurchase는 로또 구매 수량이랑 구매 금액 정보를 따로 validation 체크해서 UserOrder에 포함시키고 싶어서 그렇게 한건데, 그냥 객체로 분리하지 말고 UserOrder에 변수로 두는 것이 더 나을까요??

limwoobin replied:
앗 ㅎㅎ 구매수량과 금액정보를 나타내려는 의도이셨군요 
그렇다면 지금방향으로 한번 진행해도 좋을것같네요! 
다만 UserPurchase의 인스턴스 변수로 있는 amount에 대해서는 인스턴스변수로 존재할 필요가 있을까? 에 대해서는
한번 고민해보시면 좋을거같아요~
외부에서 사용되는부분은 없는걸로 보여서요 ㅎㅎ

jisu1211 replied:
네 감사합니다! ㅎㅎ

limwoobin commented on src/main/java/study/lotto/domain/order/AutoOrder.java:
```diff
@@ -0,0 +1,25 @@
+
+import java.util.List;
+
+public class AutoOrder implements Order {
```
자동구매와 수동구매에 대해 객체로 래핑해주셨군요! 
굉장히 세세하게 객체를 나눠주셨네요 인상적입니다 ㅎㅎ

limwoobin replied:
하지만 해당 객체에 대해서는 단위 테스트가 빠져있네요
확인한번 부탁드립니다!

limwoobin commented:
추가적으로 이번 요구사항에는 `예외 처리를 통해 에러가 발생하지 않도록 한다.` 와 같은 요구사항이 있었습니다.
입력값에 대해 Exception을 리턴하는것이 아닌 예외처리를 통해 해당 입력값을 다시 retry 할 수 있도록 요구사항을
반영해보시는건 어떨까요??

limwoobin commented:
이번 미션에서는 view 영역과 domain 영역의 의존성을 분리해보는것도 지수님께 많은 공부가 될것같습니다!
예를 들면 `LottoInput` 의 경우 `WinningLotto`와 같은 도메인객체를 알고있네요 

이해가 안가신다거나 고민되시는 부분이 있으면 편하게 dm 주세요!

limwoobin reviewed (changes requested):
안녕하세요 지수님 
드디어 마지막 단계네요.
마지막 단계인만큼 이대로 머지하기엔 아쉬운 부분이 있어 코멘트 드렸습니다.

다시한번 확인하시고 리뷰요청 부탁드립니다!

jisu1211 commented:
> toLottoNumber가 String, Integer 두 타입을 모두 받고있네요 ㅎㅎ
> 이 두 방식을 하나의 인자만 받게끔 제한시켜 toLottoNumber를 하나로 사용하는건 어떨까요?

-> `toLottoNumber(int number)` 로 일원화하여 수정했습니다!

> LottoGenerator로 부터 아에 로또객체를 리턴받는게
> 더 객체의 역할에 알맞지 않을까요??

-> `Lotto generate(List<Integer> numbers, OrderType orderType)` 로또 객체를 생성하여 넘기도록 수정했습니다!

> userOrder라는 곳에서 Lottos를 리턴하는군요
> Store라는 객체가 있는데 userOrder에서 로또를 리턴하는게 어색하네요
> 아에 Store에게 UserOrder와 같은 객체를 전달해서 리턴받는 방법도 있을것같습니다.

-> `Lottos buyLottos(Order order)` 이렇게 수정해봤습니다!

> 이런 메서드는 view영역으로 이동시키는게 어떨까요??

-> view 영역으로 이동했습니다!

> 앗 ㅎㅎ 구매수량과 금액정보를 나타내려는 의도이셨군요
> 그렇다면 지금방향으로 한번 진행해도 좋을것같네요!
> 다만 UserPurchase의 인스턴스 변수로 있는 amount에 대해서는 인스턴스변수로 존재할 필요가 있을까? 에 대해서는
> 한번 고민해보시면 좋을거같아요~
> 외부에서 사용되는부분은 없는걸로 보여서요 ㅎㅎ

-> 우빈님 말씀대로, 굳이 amount를 변수로 유지할 필요가 없을 것 같아서 `Quantity` 객체를 생성하는 방향으로 수정해봤습니다!

> 이번 미션에서는 view 영역과 domain 영역의 의존성을 분리해보는것도 지수님께 많은 공부가 될것같습니다!
> 예를 들면 LottoInput 의 경우 WinningLotto와 같은 도메인객체를 알고있네요
>
> 이해가 안가신다거나 고민되시는 부분이 있으면 편하게 dm 주세요!

-> view와 domain 영역을 나름 `LottoService`, `OrderService`를 생성해서 분리해봤는데 이 부분 어떻게 생각하시는지 리뷰 부탁드립니다! 어렵네요 ㅠ

> 추가적으로 이번 요구사항에는 예외 처리를 통해 에러가 발생하지 않도록 한다. 와 같은 요구사항이 있었습니다.
> 입력값에 대해 Exception을 리턴하는것이 아닌 예외처리를 통해 해당 입력값을 다시 retry 할 수 있도록 요구사항을
> 반영해보시는건 어떨까요??

-> 제가 요구사항을 놓쳤었네요 ㅠㅠ 입력되는 값별로 다시 retry되도록 수정했습니다!

jisu1211 committed:
refactor: Order 관련 클래스 삭제

jisu1211 committed:
refactor: view와 controller, domain 분리

jisu1211 committed:
refactor: OrderType 추가
test: OrderTypeTest 추가

jisu1211 committed:
docs: README.md 수정

jisu1211 committed:
refactor: Quantity 추가
- Quantity
   - 사용자의 로또 구매 수량을 관리하는 클래스.
   - 로또의 구매 금액을 입력받아 로또 구매 수량을 계산한다.
   - 로또의 구매 금액은 로또 기본 가격(1000원)보다 커야한다. -> 작은 경우 IllegalArgumentException 발생
   - 로또의 수동 구매량을 입력받아 총 구매량과 비교해 수동 구매할 수 있는지 확인한다.
   - 로또의 수동 구매량을 입력받아 자동 구매량을 계산할 수 있다.
test: OrderTest, QuantityTest 추가

jisu1211 committed:
refactor: LottoGenerator 리팩터링
- toLottoNumber 메서드 일원화
- generate 메서드를 추가해서 Lotto 객체 리턴하도록 수정

test: LottoGeneratorTest 수정

jisu1211 committed:
refector: 기존 LottoGeneratorTest 삭제

jisu1211 committed:
refactor: Store 리팩터링
- 사용자 주문 정보를 입력받아 Lottos를 생성할 수 있는 메서드 추가
- 사용자가 수동/자동으로 나누어서 주문할 수 있도록 수정

test: StoreTest 수정

jisu1211 committed:
feat: Lottos 기능 추가
- 주문 타입별 개수를 계산하는 메서드 추가
- 주문 타입이 수동인 로또의 개수를 계산하는 메서드 추가
- 주문 타입이 자동인 로또의 개수를 계산하는 메서드 추가

test: LottosTest 수정

jisu1211 committed:
refactor: Lotto 리팩터링
- List를 통해서 로또 번호를 입력받는 생성자 제거

feat: Lotto 기능 추가
- OrderType을 생성자로 입력받도록 수정
- 현재 로또가 수동으로 구매된 로또인지 확인하는 메서드 추가
- 현재 로또가 자동으로 구매된 로또인지 확인하는 메서드 추가

test: LottoTest 수정

jisu1211 committed:
refactor: 메서드명 수정, 사용하는 메서드 변경
test: WinningLottoTest, WinStatsTest 수정

jisu1211 committed:
refactor: LottoGames에서 domain과의 의존성 제거
- LottoService 추가
  - 하나의 로또 목록 정보, 지난 주 로또 정보, 추첨 결과 정보를 가지며 외부에서 들어오는 로또 관련 요청을 처리한다.
- OrderService 추가
  - 하나의 주문을 가지며 외부에서 들어오는 주문 관련 요청을 처리한다.

test: LottoServiceTest, OrderServiceTest 추가

limwoobin commented on src/main/java/study/lotto/domain/order/OrderType.java:
```diff
@@ -0,0 +1,13 @@
+package study.lotto.domain.order;
+
+public enum OrderType {
```
enum을 이용하셨군요 ㅎㅎ
전보다 수동,자동에 대한 구분도 명확하고 관련메소드를 나타낼 수 있어서 좋아보이네요!

jisu1211 replied:
감사합니다~ 우빈님!
피드백으로 생각지도 못한 부분이나 더 개선하면 좋을 점들을 많이 짚어주셔서 
개선해나가는 과정에서 어렵기도 했지만 그만큼 발전한 것 같아서 좋습니다 ㅎㅎ
코드를 작성할 때 제 나쁜 버릇도 많이 발견했네요.. ㅎㅎ 😢
앞으로 열심히 연습해야겠습니다!
지금까지 좋은 피드백 주셔서 감사합니다 😊

좋은 주말 보내세요!

limwoobin commented on README.md:
```diff
@@ -113,4 +113,61 @@
+    - 주문에 수동 구매량을 입력한다.
+    - 주문에 수동 구매 로또들을 입력한다.
+    - 수동 구매를 더 할 수 있는지 확인한다.
+  
```
꼼꼼하게 클래스의 역할까지 기재해서 작성해주셨네요 
저도 많이 배워야겠습니다 ㅎㅎ

limwoobin commented on src/main/java/study/lotto/domain/number/LottoGenerator.java:
```diff
@@ -20,15 +23,13 @@ public class CacheLottoNumbers {
+        return new Lotto(numbers.stream()
+                .map(LottoGenerator::toLottoNumber)
+                .collect(Collectors.toSet()), orderType);
     }
```
Lotto를 리턴하는 로직으로 변경되니 
전보다 훨씬 자연스러워보입니다! 객체의 역할도 명확해졌네요 ㅎㅎ

limwoobin commented on src/test/java/study/lotto/domain/StoreTest.java:
```diff
@@ -1,22 +1,69 @@
+            Store.buyLottoManually(lotto);
+        }).isInstanceOf(IllegalArgumentException.class)
+                .hasMessage(LottoExceptionCode.INVALID_LOTTO_NUMBER.getMessage());
+    }
```
예외에 대한 테스트까지 꼼꼼하게 작성해주셨군요 ㅎㅎ

limwoobin commented on src/main/java/study/lotto/LottoGames.java:
```diff
@@ -1,61 +1,92 @@
+            output.printMessage(iae.getMessage());
+            inputAmount();
+        }
     }
```
도메인과의 분리도 성공적으로 잘해주셨습니다!
한가지 의견을 드리자면 `output.printAmountMessage();` 와 같은 기능은
`LottoInput` 에서 가지고있는게 어떨까 합니다 
Input, Output의 구분이 입력받는 동작을 하느냐, 출력받는 동작을 하느냐인데
단순 기능으로 구분하기 보다는 입력받는 동작에는 입력을 위한 메시지가 포함되어도 된다고 생각해서 그런데요ㅎㅎ 지수님 의견은 어떠신가요??

jisu1211 replied:
우빈님 말씀대로 저 부분을 작성할 때 입력과 출력을 나누기가 애매한 부분이 있었습니다,,
저는 일단은 출력이기 때문에 LottoOutput에 작성했는데 
우빈님이 제안해주신대로 연관성이 있는 작업이다보니 그렇게 작성해도 좋을 것 같습니다!
마지막까지 좋은 의견 감사해요~! ㅎㅎ

limwoobin reviewed (approved):
안녕하세요 지수님 
드디어 로또 미션도 끝이났네요 ㅎㅎ
마지막까지 고민하신 흔적이 많이 느껴져서 정말 좋았습니다 👍 
저도 많이 배웠습니다 ㅎㅎ

소소한 코멘트를 드렸으니 한번 확인부탁드려요
마지막까지 정말 고생많으셨습니다. 이번 미션으로 제가 드린 피드백이
지수님께 조금이라도 도움이되었으면 합니다 ㅎㅎ

저와의 미션은 여기서 끝이지만 이후 미션을 진행하며 생기는 궁금증이나 의논하고싶은점이 있다면
편하게 dm 주셔서 의견주셔도 좋아요 ㅎㅎ
그럼 다음 미션도 화이팅입니다!

limwoobin commented:
> > toLottoNumber가 String, Integer 두 타입을 모두 받고있네요 ㅎㅎ
> > 이 두 방식을 하나의 인자만 받게끔 제한시켜 toLottoNumber를 하나로 사용하는건 어떨까요?
> 
> -> `toLottoNumber(int number)` 로 일원화하여 수정했습니다!
> 
> > LottoGenerator로 부터 아에 로또객체를 리턴받는게
> > 더 객체의 역할에 알맞지 않을까요??
> 
> -> `Lotto generate(List<Integer> numbers, OrderType orderType)` 로또 객체를 생성하여 넘기도록 수정했습니다!
> 
> > userOrder라는 곳에서 Lottos를 리턴하는군요
> > Store라는 객체가 있는데 userOrder에서 로또를 리턴하는게 어색하네요
> > 아에 Store에게 UserOrder와 같은 객체를 전달해서 리턴받는 방법도 있을것같습니다.
> 
> -> `Lottos buyLottos(Order order)` 이렇게 수정해봤습니다!
> 
> > 이런 메서드는 view영역으로 이동시키는게 어떨까요??
> 
> -> view 영역으로 이동했습니다!
> 
> > 앗 ㅎㅎ 구매수량과 금액정보를 나타내려는 의도이셨군요
> > 그렇다면 지금방향으로 한번 진행해도 좋을것같네요!
> > 다만 UserPurchase의 인스턴스 변수로 있는 amount에 대해서는 인스턴스변수로 존재할 필요가 있을까? 에 대해서는
> > 한번 고민해보시면 좋을거같아요~
> > 외부에서 사용되는부분은 없는걸로 보여서요 ㅎㅎ
> 
> -> 우빈님 말씀대로, 굳이 amount를 변수로 유지할 필요가 없을 것 같아서 `Quantity` 객체를 생성하는 방향으로 수정해봤습니다!
> 
> > 이번 미션에서는 view 영역과 domain 영역의 의존성을 분리해보는것도 지수님께 많은 공부가 될것같습니다!
> > 예를 들면 LottoInput 의 경우 WinningLotto와 같은 도메인객체를 알고있네요
> > 이해가 안가신다거나 고민되시는 부분이 있으면 편하게 dm 주세요!
> 
> -> view와 domain 영역을 나름 `LottoService`, `OrderService`를 생성해서 분리해봤는데 이 부분 어떻게 생각하시는지 리뷰 부탁드립니다! 어렵네요 ㅠ
> 
> > 추가적으로 이번 요구사항에는 예외 처리를 통해 에러가 발생하지 않도록 한다. 와 같은 요구사항이 있었습니다.
> > 입력값에 대해 Exception을 리턴하는것이 아닌 예외처리를 통해 해당 입력값을 다시 retry 할 수 있도록 요구사항을
> > 반영해보시는건 어떨까요??
> 
> -> 제가 요구사항을 놓쳤었네요 ㅠㅠ 입력되는 값별로 다시 retry되도록 수정했습니다!

LottoService, OrderService와 로 분리하신건 굉장히 좋다고 생각합니다!
물론 해당 로직을 `LottoGames`에서 처리해도 충분할 수 있지만 저희의 이번 미션은 
책임에 맞게 객체를 분리하는 시도를 하며 객체지향을 공부하는 과정이니깐요 ㅎㅎ
그리고 실제로 역할에 맞게 분리도 잘해주셨습니다! 추후 로직이 커지면 위와같은 분리는 필요할 수 있는데 이런 부분들을 잘 적용해주셨어요 👍
